v3/src/models/history/apply-model-change.ts

```diff
@@ -1,69 +1,29 @@
 import { ILogMessage } from "../../lib/log-message"
-import { getTileEnvironment } from "../tiles/tile-environment"
-import { withUndoRedoStrings } from "./codap-undo-types"
-import { withoutUndo } from "./without-undo"
+import { getHistoryService } from "./history-service"
 
-export interface INotification {
-  message: DIMessage
-  callback?: iframePhone.ListenerCallback
-}
 export type INotify =
-  INotification | (INotification | (() => Maybe<INotification>))[] | (() => Maybe<INotification | INotification[]>)
+  ICoreNotification |
+  (ICoreNotification | (() => Maybe<ICoreNotification>))[] |
+  (() => Maybe<ICoreNotification | ICoreNotification[]>)
 export interface IApplyModelChangeOptions {
   log?: string | ILogMessage | (() => Maybe<string | ILogMessage>)
   notify?: INotify
   notifyTileId?: string
   undoStringKey?: string
   redoStringKey?: string
 }
+
 // returns an object which defines the `applyModelChange` method on an MST model
 // designed to be passed to `.actions()`, i.e. `.actions(applyModelChange)`
 export function applyModelChange(self: IAnyStateTreeNode) {
   return ({
     // performs the specified action so that response actions are included and undo/redo strings assigned
     applyModelChange<TResult = unknown>(actionFn: () => TResult, options?: IApplyModelChangeOptions) {
-      const { log, notify, notifyTileId, undoStringKey, redoStringKey } = options || {}
       const result = actionFn()
 
-      // Add strings to undoable action or keep out of the undo stack
-      if (undoStringKey != null && redoStringKey != null) {
-        withUndoRedoStrings(undoStringKey, redoStringKey)
-      } else {
-        withoutUndo()
-      }
-
-      const tileEnv = getTileEnvironment(self)
-      if (tileEnv) {
-        // Send log message to logger
-        if (tileEnv.log) {
-          const logInfo = typeof log === "function" ? log() : log
-          const logMessage = typeof logInfo === "string" ? { message: logInfo } : logInfo
-          if (logMessage) {
-            tileEnv.log(logMessage)
-          }
-        }
-
-        // Broadcast notifications to plugins
-        if (notify && tileEnv.notify) {
-          // Convert notifications to INotification[]
-          const actualNotifications = notify instanceof Function ? notify() : notify
-          if (actualNotifications) {
-            const notificationArray = Array.isArray(actualNotifications) ? actualNotifications : [actualNotifications]
-
-            // Actually broadcast the notifications
-            notificationArray.forEach(_notification => {
-              const __notification = typeof _notification === "function" ? _notification() : _notification
-              if (__notification) {
-                const { message, callback } = __notification
-                tileEnv.notify?.(message, callback ?? (() => null), notifyTileId)
-              }
-            })
-          }
-        }
-      }
+      getHistoryService().handleApplyModelChange(self, options)
```

Comment:
Couldn't you also attach the history service to the MST environment like we do with logging, which would avoid the need to register the history service explicitly for tests that only require it for `applyModelChange`? You probably considered that and rejected it for some reason, but it seems like it would have some advantages.

Reply:
I think I ruled it out because I was trying to avoid depending on the `ITileEnvironment`. However as I've continued on this work I now think using ITileEnvironment makes sense. 

I also was thinking about not changing all of the code that calls `withoutUndo()`. But I've figure out a way to look up the MST environment from a simple `withoutUndo()` so I think its do-able. I'm running out of time on this, but hopefully I can get it wrapped up.